src/utils/special-paths.js

```diff
@@ -119,3 +126,76 @@ export function parseFileNameFromSymbolication(
     path: file,
   };
 }
+
+// Find out where the raw source code for this file can be obtained.
+// Since we want to fetch the source code with the browser, the server needs
+// to send permissive CORS headers.
+// Some files cannot be obtained from the internet at all. Others can only be
+// obtained as part of an archive.
+export function getDownloadRecipeForSourceFile(
+  parsedFile: ParsedFileNameFromSymbolication
+): SourceFileDownloadRecipe {
+  switch (parsedFile.type) {
+    case 'hg': {
+      // Assume that there is a hgweb instance running on the server.
+      // It serves raw files with permissive CORS headers at /raw-file/.
+      // This is true at least for hg.mozilla.org.
+      const { repo, rev, path } = parsedFile;
+      return {
+        type: 'CORS_ENABLED_SINGLE_FILE',
+        url: `https://${repo}/raw-file/${rev}/${path}`,
+      };
+    }
+    case 'git': {
+      // For git, special-case the URL based on the repository host.
+      const { repo, rev, path } = parsedFile;
+      // Example repo strings:
+      // "github.com/rust-lang/rust"
+      // "chromium.googlesource.com/chromium/src"
+      const slashPos = repo.indexOf('/');
+      if (slashPos === -1) {
+        return { type: 'NO_KNOWN_CORS_URL' };
+      }
+      const server = repo.slice(0, slashPos);
+      const repoPath = repo.slice(slashPos + 1);
+      if (server === 'github.com') {
+        return {
+          type: 'CORS_ENABLED_SINGLE_FILE',
+          url: `https://raw.githubusercontent.com/${repoPath}/${rev}/${path}`,
+        };
+      }
+      if (server.endsWith('.googlesource.com')) {
+        const subdomain = server.slice(
+          0,
+          server.length - '.googlesource.com'.length
+        );
+        return {
+          type: 'CORS_ENABLED_SINGLE_FILE',
+          url: `https://googlesource-proxy.mstange.workers.dev/${subdomain}/${repoPath}.git/+/${rev}/${path}`,
```

Comment:
Oh, I didn't know about this URL. Should this be moved to a nicer domain at some point? And is the source code somewhere we can see?

Reply:
Heh, nobody knows about this URL :)

It's a free Cloudflare worker for now, and I fear that if the URL becomes more widely-known it might no longer be free. I've put the source code [here](https://gist.github.com/mstange/ed6f7e494b763abc755f3d6e753924ee).